###################################################################################################
Date: Fri, 7 Dec 2018 14:53:31 +0000
From 35484ddb1b3d966e2e25d4f27d62206adabf326d Mon Sep 17 00:00:00 2001
From: Ben Gray <ben.gray@sky.uk>

Subject: [PATCH] Added a config option to enable / disable LE secure connections.
Useful for sky as some of our RCUs don't support SC and incorrectly report that they do support SC

Source: SKY
Upstream-Status: Pending
Signed-off-by: Tijo Thomas John <tijo.thomas@sky.uk>
###################################################################################################

---
 src/adapter.c | 10 ++++++++--
 src/hcid.h    |  1 +
 src/main.c    | 11 +++++++++++
 src/main.conf |  5 +++++
 4 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/src/adapter.c b/src/adapter.c
index 913fde1..d6e1ac7 100644
--- a/src/adapter.c
+++ b/src/adapter.c
@@ -8305,8 +8305,14 @@ static void read_info_complete(uint8_t status, uint16_t length,
 		break;
 	}
 
-	if (missing_settings & MGMT_SETTING_SECURE_CONN)
-		set_mode(adapter, MGMT_OP_SET_SECURE_CONN, 0x01);
+	if (adapter->supported_settings & MGMT_SETTING_SECURE_CONN) {
+		if (main_opts.secure_conn &&
+				!(adapter->current_settings & MGMT_SETTING_SECURE_CONN))
+			set_mode(adapter, MGMT_OP_SET_SECURE_CONN, 0x01);
+		else if (!main_opts.secure_conn &&
+				(adapter->current_settings & MGMT_SETTING_SECURE_CONN))
+			set_mode(adapter, MGMT_OP_SET_SECURE_CONN, 0x00);
+	}
 
 	if (adapter->supported_settings & MGMT_SETTING_PRIVACY)
 		set_privacy(adapter, main_opts.privacy);
diff --git a/src/hcid.h b/src/hcid.h
index ea8d60c..74ad583 100644
--- a/src/hcid.h
+++ b/src/hcid.h
@@ -56,6 +56,7 @@ struct main_opts {
 	bt_gatt_cache_t gatt_cache;
 
 	gboolean	kern_conn_control;
+	gboolean	secure_conn;
 };
 
 extern struct main_opts main_opts;
diff --git a/src/main.c b/src/main.c
index ee0a74c..6a85159 100644
--- a/src/main.c
+++ b/src/main.c
@@ -110,6 +110,7 @@ static const char *gatt_options[] = {
 
 static const char *sky_options[] = {
 	"KernelConnectionControl",
+	"SecureConnections",
 	NULL
 };
 
@@ -309,6 +310,15 @@ static void parse_sky_config(GKeyFile *config)
 	else
 		main_opts.kern_conn_control = boolean;
 
+	boolean = g_key_file_get_boolean(config, "sky",
+						"SecureConnections", &err);
+	if (err) {
+		g_clear_error(&err);
+	} else {
+		DBG("SecureConnections=%s", boolean ? "true" : "false");
+		main_opts.secure_conn = boolean;
+	}
+
 }
 
 static void parse_config(GKeyFile *config)
@@ -485,6 +495,7 @@ static void init_defaults(void)
 
 	/* Default settings for Sky extensions */
 	main_opts.kern_conn_control = TRUE;
+	main_opts.secure_conn = TRUE;
 }
 
 static void log_handler(const gchar *log_domain, GLogLevelFlags log_level,
diff --git a/src/main.conf b/src/main.conf
index 878592e..9efcd7c 100644
--- a/src/main.conf
+++ b/src/main.conf
@@ -108,3 +108,8 @@
 # ourselves. Defaults to 'true'.
 # KernelConnectionControl = true
 
+# Enables or disables bluetooth LE "Secure Connections".  This is a BT 4.1
+# feature used at pairing time, ideally it should be enabled but can cause
+# interopt issues with some BLE devices. Defaults to 'true'.
+# SecureConnections = true
+
-- 
2.18.0

